### PR TITLE
`<deque>`: Fix iterator arithmetic

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Currently we sometimes directly add a `size_type` value to a fancy pointer, which is not always supported. Also, the difference type of `pointer` and `_Mapptr` may be different (which is contrived though). I think we should `static_cast` to the correct difference type in iterator arithmetic.

The return value of `_Getblock` is always used in the subscipt, so I guess it's OK to change the return type to the difference type.

Driven-by: avoid seemingly redundant indirection in `operator[]` and `at`. (Perhaps we should also do this for `front` and `back`).

### CODE REVIEW

File: stl/inc/deque
Comment (achabense) [inline stl/inc/deque:563]: I think for consistency with the context, the parameter would better be renamed `_Off`.
And for less confusion, I'd suggest renaming this method(`_Deque_val::_At(map-index)`), and defining a `deque::_At(public-index)` instead.

**Comment by achabense** [comment]
Hmm, I think it will be horrible to have two `difference_type`s (`difference_type` and `_Map_difference_type`, from different sources) here and there :|

File: stl/inc/deque
Comment (achabense) [inline stl/inc/deque:560]: I think we should put off every `cast-to-_Map_difference_type` to caller site, *only when having to add to a pointer*.

File: stl/inc/deque
Comment (achabense) [inline stl/inc/deque:565]: If I'm correct, this `difference_type` come from `_Alty_traits::difference_type`, which has a different source from `_Map_difference_type`.
My suggestions are:
1. not changing return type of `_Getblock`.
2. this function should be
```
const size_type _Block = _Getblock(_Idx);
const auto _Off            = static_cast<_Map_difference_type >(_Idx % _Block_size);
return _Map[_Block][_Off];
```

File: stl/inc/deque
Comment (achabense) [inline stl/inc/deque:74]: As a side note, these `different_type` comes from `_Alty_traits`.

File: stl/inc/deque
Comment (achabense) [inline stl/inc/deque:1691]: this should belong to `_Deque_val`, as it's taking `_Idx` as map-index instead of public-index.

File: stl/inc/deque
Comment (frederick-vs-ja) [reply stl/inc/deque:565]: > If I'm correct, this `difference_type` comes from `_Alty_traits::difference_type`, which has a different source from `_Map_difference_type` (`_Al[p]ty_traits`).

This is right.

> My suggestions are:
> 
> 1. not changing return type of `_Getblock`.
> 2. this function should be
> 
> ```
> const size_type _Block = _Getblock(_Idx);
> const auto _Off        = static_cast<_Map_difference_type>(_Idx(or "_Off") % _Block_size);
> return _Map[_Block][_Off];
> ```

These suggestions don't seem correct. The new test case will be rejected with these applied.

Comment (achabense) [reply stl/inc/deque]: > The new test case will be rejected with these applied.

Yes, I mistook the types here :(

File: stl/inc/deque
Comment (achabense) [inline stl/inc/deque:1559]: It might be better for clarity to replace `_Map_off` and `_Map_count` variables with on-site `static_cast`s here.

I have no idea about whether it's worthwhile to replace `_Map_distance` with on-site `static_cast`s, though.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! I think this needs only one maintainer approval.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for improving our fancy pointer support! :monocle_face: :tophat: :tada:

File: stl/inc/deque
Comment (frederick-vs-ja) [reply stl/inc/deque]: I'm adding `deque::_At` in new PR #4071.

**Pull request merged**
